sycl/doc/Assert.md

```diff
@@ -0,0 +1,269 @@
+# Assert feature
+
+**IMPORTANT**: This document is a draft.
+
+Using the standard C++ `assert` API ("assertions") is an important debugging
+technique widely used by developers. This document describes the design of
+supporting assertions within SYCL device code.
+The basic approach we chose is delivering device-side assertions as host-side
+asynchronous exceptions, which allows further extensibility, such as better
+error handling or potential recovery.
+
+As usual, device-side assertions can be disabled by defining `NDEBUG` macro at
+compile time.
+
+## Use-case example
+
+```
+using namespace sycl;
+auto ErrorHandler = [] (exception_list Exs) {
+  for (exception_ptr const& E : Exs) {
+    try {
+      std::rethrow_exception(E);
+    }
+    catch (event_error const& Ex) {
+      std::cout << “Exception - ” << Ex.what(); // assertion failed
+      std::abort();
+    }
+  }
+};
+
+void user_func(item<2> Item) {
+  assert((Item[0] % 2) && “Nil”);
+}
+
+int main() {
+  queue Q(ErrorHandler);
+  q.submit([&] (handler& CGH) {
+    CGH.parallel_for<class TheKernel>(range<2>{N, M}, [=](item<2> It) {
+      do_smth();
+      user_func(It);
+      do_smth_else();
+    });
+  });
+  Q.wait_and_throw();
+  std::cout << “One shouldn’t see this message.“;
+  return 0;
+}
+```
+
+In this use-case every work-item with even X dimension will trigger assertion
+failure. Assertion failure should be reported via asynchronous exceptions. If
+asynchronous exception handler is set the failure is reported with
+`sycl::event_error` exception. Otherwise, SYCL Runtime should trigger abort.
+Even though multiple failures of the same or different assertions can happen in
+multiple workitems, implementation is required to deliver only one. The
+assertion failure message is printed to `stderr` by SYCL Runtime.
```

Comment:
Does it happen always or only without `async_handler` set?

Reply:
I think it should always print the assertion message because:

* This would be consistent with the "safe" implementation (the one that depends on hardware support), which is defined to print the message even before notifying the host.

* This is also consistent with the way `assert` works on the host, which prints the assertion message even before raising `SIGABRT`.

Reply:
So, even if user set an `async_handler` in order to gracefully react to `assert` failure, we still print something to `stderr`? What for? It is not that bad as if we printed into `stdout`, but still seems unnecessary a bit.

Reply:
I thought it was weird when I first read this spec also.  But then I tried the following test:

```
#include <cassert>
#include <csignal>
#include <cstdlib>

void handle(int sig) {
  std::exit(0);  // Exit silently
}

int main() {
  std::signal(SIGABRT, handle);
  assert(false);
}
```

The results:

```
$ clang -std=c++17 -pedantic -o t t.cpp
$ ./t
t: t.cpp:11: int main(): Assertion `false' failed.
$ echo $?
0
```

Despite the fact that I catch the SIGABRT and exit without printing anything, I still get a message printed to stderr.

Therefore, it seems like the behavior defined in this spec is consistent with the way `assert` works on the host.

Reply:
That's kind of obvious due to the fact that `assert(expr)` on host is unwrapped into
```
if (!(expr)) {
  fprintf(stderr, ...);
  abort();
}
```

In device-code, `assert(expr)` unwraps to:
```
if (!(expr)) {
  __devicelib_assert_fail(#expr, __FILE__, __LINE__, __PRETTY_FUNCTION__, global ID, local ID);
}
```